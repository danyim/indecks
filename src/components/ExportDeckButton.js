import React from 'react';

class ExportDeckButton extends React.Component {
  constructor(props) {
    super(props);
  }

  handleDownloadClick(event) {
    function magicDownload(text, fileName){
      var blob = new Blob([text], {
        type: 'text/csv;charset=utf8;'
      });

      // create hidden link
      var element = document.createElement('a');
      document.body.appendChild(element);
      element.setAttribute('href', window.URL.createObjectURL(blob));
      element.setAttribute('download', fileName);
      element.style.display = '';

      element.click();

      document.body.removeChild(element);
      event.stopPropagation();
    }

    var fileType = event.target.innerText,
    text = this.props.exportFile(fileType)

    if (text instanceof Promise) {
      text.then(
        result => magicDownload(result, this.props.filename)
      )
    } else {
      magicDownload(text, this.props.filename)
    }

  }

  render() {
    return (
      <button style={ this.props.style }
        className={ this.props.className }
        onClick={ this.handleDownloadClick }>
        {this.props.label}
      </button>
    );
  }
}

ExportDeckButton.propTypes = {
  filename: React.PropTypes.string,
  label: React.PropTypes.string,
  style: React.PropTypes.object,
  className: React.PropTypes.string,
  // exportFile: React.PropTypes.function,
};


ExportDeckButton.defaultProps = {
  filename: 'file.txt',
  label: 'Save',
  style: { margin: '5px 5px 0px 0px', textDecoration: 'underline', color: 'blue', cursor: 'pointer' },
  exportFile: () => {}
};

export default ExportDeckButton
