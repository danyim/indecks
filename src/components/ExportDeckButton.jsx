import React from 'react';

const propTypes = {
  filename: React.PropTypes.string,
  label: React.PropTypes.string,
  disabled: React.PropTypes.bool,
  style: React.PropTypes.object,
  className: React.PropTypes.string,
  exportFile: React.PropTypes.func
};

const defaultProps = {
  filename: 'file.txt',
  label: 'Save',
  disabled: false,
  style: {
    margin: '5px 5px 0px 0px',
    textDecoration: 'underline',
    color: 'blue',
    cursor: 'pointer'
  },
  exportFile: () => {}
};

class ExportDeckButton extends React.Component {
  constructor(props) {
    super(props);

    this.handleDownloadClick = this.handleDownloadClick.bind(this);
  }

  magicDownload(text, fileName) {
    const blob = new Blob([text], {
      type: 'text/csv;charset=utf8;'
    });

    // create hidden link
    const element = document.createElement('a');
    document.body.appendChild(element);
    element.setAttribute('href', window.URL.createObjectURL(blob));
    element.setAttribute('download', fileName);
    element.style.display = '';

    element.click();

    document.body.removeChild(element);
    event.stopPropagation();
  }

  handleDownloadClick(event) {
    const fileType = event.target.innerText;
    const text = this.props.exportFile(fileType);

    if (text instanceof Promise) {
      text.then(
        (result) => this.magicDownload(result, this.props.filename)
      );
    } else {
      this.magicDownload(text, this.props.filename);
    }
  }

  render() {
    return (
      <button
        style={this.props.style}
        className={this.props.className}
        onClick={this.handleDownloadClick}
        disabled={this.props.disabled}
      >
        {this.props.label}
      </button>
    );
  }
}

ExportDeckButton.propTypes = propTypes;
ExportDeckButton.defaultProps = defaultProps;

export default ExportDeckButton;
