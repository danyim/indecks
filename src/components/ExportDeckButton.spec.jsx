import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import ExportDeckButton from './ExportDeckButton'

const defaultProps = {
  filename: 'file.txt',
  label: 'Save',
  className: '',
  disabled: false,
  style: {
    margin: '5px 5px 0px 0px',
    textDecoration: 'underline',
    color: 'blue',
    cursor: 'pointer',
  },
  exportFile: () => {},
}

function setup(props = defaultProps) {
  const wrapper = shallow(<ExportDeckButton {...props} />)

  return {
    props,
    wrapper,
  }
}

describe('ExportDeckButton', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup()
    expect(wrapper.find('button').hasClass(defaultProps.className)).toBe(true)

    const tree = renderer
      .create(<ExportDeckButton {...defaultProps} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  // it('should call the exportFile once when ExportDeckButton is clicked', () => {
  //   const handler = jest.fn();
  //   const { wrapper } = setup({
  //     ...defaultProps,
  //     exportFile: handler
  //   });

  //   wrapper.find('button').simulate('click');
  //   expect(handler.mock.calls.length).toBe(1);
  // });
})
