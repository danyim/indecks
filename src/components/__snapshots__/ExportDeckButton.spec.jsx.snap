// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportDeckButton should render self and subcomponents 1`] = `
<button
  className=""
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "color": "blue",
      "cursor": "pointer",
      "margin": "5px 5px 0px 0px",
      "textDecoration": "underline",
    }
  }
>
  Save
</button>
`
